packages/datatrak-web/src/sync/ClientSyncManager.ts

```diff
@@ -0,0 +1,110 @@
+import { modelClasses, ModelRegistry } from '@tupaia/database';
+import {
+  FACT_LAST_SUCCESSFUL_SYNC_PULL,
+  SYNC_DIRECTIONS,
+  createClientSnapshotTable,
+  dropAllSnapshotTables,
+  dropSnapshotTable,
+  getModelsForDirection,
+  saveIncomingInMemoryChanges,
+  saveIncomingSnapshotChanges,
+} from '@tupaia/sync';
+import { DatatrakDatabase } from '../database/DatatrakDatabase';
+import { initiatePull, pullIncomingChanges } from './pullIncomingChanges';
+import {
+  getPullVolumeType,
+  PROCESS_STREAM_DATA_FUNCTIONS,
+  PullVolumeType,
+} from './processStreamedData';
+import { post, remove } from '../api';
+import { DatatrakWebModelRegistry } from '../types/model';
+
+export interface SyncResult {
+  ran: boolean;
+}
+
+export class ClientSyncManager {
+  private database: DatatrakDatabase;
+
+  private models: DatatrakWebModelRegistry;
+
+  private currentSyncPromise: Promise<SyncResult> | null = null;
+
+  private projectIds: string[];
+
+  private deviceId: string;
+
+  constructor(models: DatatrakWebModelRegistry, projectIds: string[], deviceId: string) {
+    this.models = models;
+    this.database = models.database;
+    this.projectIds = projectIds;
+    this.deviceId = deviceId;
+  }
+
+  async triggerSync() {
+    if (this.currentSyncPromise) {
+      return this.currentSyncPromise;
+    }
+
+    // set up a common sync promise to avoid double sync
+    this.currentSyncPromise = this.runSync();
+
+    // make sure sync promise gets cleared when finished, even if there's an error
+    try {
+      return this.currentSyncPromise;
+    } finally {
+      this.currentSyncPromise = null;
+    }
```

Comment:
Why are we immediately setting the currentSyncPromise to `null` here? Should this be attached to the `runSync` promise?

Reply:
That's a silly mistake of mine sorry, have fixed